Add AGENTS.md file (#333)
entur/cloud-logging: AGENTS.md

# Agents

## Project Overview

`cloud-logging` is a multi-module Gradle project providing JVM logging libraries for GCP and Azure, built on SLF4J 2, Logback, and Logback Logstash. It is **not** a generic logging library — it is specifically designed to produce structured JSON logs consumed by GCP Cloud Logging (via Fluentbit) and Azure Monitor.

## Build & Test

```bash
# Build all modules
./gradlew build

# Run tests
./gradlew test

# Build a single module (example)
./gradlew :gcp:spring-boot-starter-gcp-web:build

# Check for dependency updates
./gradlew dependencyUpdates
```

## Module Structure

- `api/` – SLF4J `Logger`-wrapper (`DevOpsLogger`) with additional severity levels: `errorTellMeTomorrow`, `errorInterruptMyDinner`, `errorWakeMeUpRightNow`
- `appender/` – Core Logback async appender with MDC support and on-demand logging scope
- `gcp/` – GCP-specific Spring Boot starters, autoconfiguration, and Stackdriver/Cloud Logging encoder
- `azure/` – Azure-specific Spring Boot starters, autoconfiguration, and Azure Monitor encoder

> - `request-response/` – Logbook-based HTTP and gRPC request/response logging

Reviewer: isn't gRPC `netty-grpc`?

- `on-demand/` – Selective on-demand logging: caches log statements per request and only emits them for failed/slow/flagged requests
- `trace/` – Correlation-id and MDC tracing for HTTP and gRPC
- `micrometer/` – Log severity metrics (GCP and Azure variants)
- `test/` – Test utilities: Logback test configuration, JUnit 5 extensions
- `examples/` – Runnable Spring Boot example applications (web and gRPC, GCP and Azure)
- `bom/` – Bill of Materials for consumer dependency management

The `gcp/` and `azure/` modules are **mirrors** of each other in structure and feature set. When adding a feature to one, add the equivalent to the other.

## Key Versions

> Dependency versions are managed centrally in the root `build.gradle` `ext` block. The Spring Boot BOM controls most transitive dependency versions (JUnit, Mockito, SLF4J, Logback, Jackson, Micrometer), and non-BOM dependency versions (e.g. `logbackLogstashVersion`, `logbookVersion`) are defined explicitly there. **Do not hardcode dependency versions in submodule `build.gradle` files** unless there is an intentional, documented exception; some example subprojects may declare plugin versions locally.

Reviewer: ```suggestion
Dependency versions are managed centrally in the root `build.gradle` `ext` block. The Spring Boot BOM controls most transitive dependency versions (JUnit, Mockito, SLF4J, Logback, Jackson, Micrometer), and non-BOM dependency versions (e.g. `logbackLogstashVersion`, `logbookVersion`) are defined explicitly there.  Example subprojects (examples/*) may declare plugin versions locally; library subprojects must not.
```

https://github.com/entur/ai/issues/11

## Conventions

> - Java with Spring Boot 4.x (not 3.x)

Reviewer: which java version? 17? 21? let the agent know the feature-set in use

- **Never add a `logback.xml` or `logback-spring.xml` manually** — the starters provide these via classpath resources; a second file causes undefined behaviour
- Log levels are configured via Spring properties (`logging.level.root`, `logging.level.<package>`)

## Log Format and Cloud Platform Mapping

### GCP (Google Cloud Logging / Stackdriver)

Logs are written as JSON to stdout/stderr. Fluentbit on GKE forwards them to Cloud Logging as `jsonPayload`. The encoder (`StackdriverLogstashEncoder`) maps:

- SLF4J level → `severity` field (not `level`) — recognized by Cloud Logging for log severity filtering
- `message` field — includes stack traces inline
- MDC keys → top-level JSON fields (or OpenTelemetry trace fields `logging.googleapis.com/trace`, `logging.googleapis.com/spanId` if OTel is on classpath)
- `serviceContext.service` — populated from `HOSTNAME` environment variable

**Important GCP constraints (Fluentbit quirks):**
- Duplicate field names in `jsonPayload` cause Fluentbit to drop 2–4 seconds of logs from all pods — avoid emitting the same key twice in one log event
- Log lines exceeding 256 KB are dropped — truncate large payloads

> - Avoid using reserved field names (`message`, `timestamp`) as structured argument keys

Reviewer: is this list exhaustive? if not, say so

External docs:
- [Azure Monitor Logs overview](https://learn.microsoft.com/en-us/azure/azure-monitor/logs/data-platform-logs)
- [Azure Monitor log standard columns](https://learn.microsoft.com/en-us/azure/azure-monitor/logs/log-standard-columns)
- [Application Insights structured logging](https://learn.microsoft.com/en-us/azure/azure-monitor/app/structured-logging)

## On-Demand Logging

On-demand logging is **disabled by default**. It must be explicitly enabled:

```
entur.logging.http.ondemand.enabled=true   # for web
entur.logging.grpc.ondemand.enabled=true   # for gRPC
```

It caches all log statements per request and either discards them (happy path) or emits them (failure/slow/debug path). Failure triggers are configurable: HTTP status code, log level, call duration, or a specific HTTP/gRPC header.

## Testing Patterns

- Always test against machine-readable JSON output under the hood, even when the console shows human-readable format during development
- Use `CompositeConsoleOutputControl` to toggle output mode within a test:
  ```java
  try (Closeable c = CompositeConsoleOutputControl.useHumanReadableJsonEncoder()) {
      // test code
  }
  ```

> - JUnit 5 extensions and Logback test utilities are in `test/test-logback-junit`

Reviewer: Maybe add a way to run a single test in isolation? Along the lines of `./gradlew :module:test --tests ClassName` when the models don't have to waste tokens on trial and error for our setup (I have no idea so I just pseudowrote a gradle line)
